### ISSUE TITLE
define requirements for a plugin architecture

### ISSUE THREAD

**Issue by kevinbarabash** [Khan/KaTeX#762]
There are a number of issues and PRs that would benefit from having a plugin architecture:
- Add support for Persian & Arabic alphabet in text and eastern-arabic numerals + Proposal Commit #729
- single or double dollar sign as formula delimiter #712
- mhchem extension support #50 
- Cursor #506 
- Class #507 

There are certain functions we could expose:
- `defineSymbol`
- `defineFunction`
- `defineEnvironment`

`defineSymbol` is used for adding support for new symbols or new languages, but if the new glyphs require new fonts currently isn't a way to specify metrics for the new glyphs.  We'd need to provide a way to extend the data stored in fontMetricsData.js.

`defineFunction` is used for adding new commands.  Commands that define implicit groups will be hard to deal and since there aren't that many of them we should probably ignore them for now.  A bigger problem is that any new function will also require code in buildHTML.js and buildMathML.js to generate the correct markup.

Currently, buildHTML.js and buildMathML.js are loose conglomeration of functions.  The only functions that should need to be shared are `buildGroup` and `buildExpression`.  The rest could exist in their own separate files.

I would like to propose grouping the parse node generation functions from `function.js` with functions from `buildHTML.js` and `buildMathML.js` into separate files based on group type.  As part of this, we'd expand the responsibilities of `defineFunction` to add entries to the `groupTypes` objects used by `buildHTML` and `buildMathML`.

We'll want to minimize changes to the parse tree structure once we publish a public plugin API so we should make changes before doing so.

**Comment by mbourne** [comment]
@kevinbarabash Would it also be appropriate to add support for ASCIIMath function in such a plugin architecture?

**Comment by kevinbarabash** [comment]
@mbourne is the idea to convert ASCIIMath syntax to LaTeX syntax and then have KaTeX render the result?  If you're calling KaTeX's render method manually then a plugin probably isn't necessary b/c you can pipe the output of something like https://github.com/tylerlong/asciimath-to-latex to KaTeX's render method.  If you you're looking to tie into the auto render extension then that will have to be updated to include an optional callback that can be run on each string to preprocess it before sending it to KaTeX.

**Comment by mbourne** [comment]
@kevinbarabash Yes, that is the idea. I'm currently piping output from an ASCII to LaTeX converter (similar to tylerlong's), which is fine for me. But I thought I would raise it as a plugin possibility to make it easier for others. MathJax did this after a few years of its existence.

**Comment by HosseinAgha** [comment]
@kevinbarabash I investigated some of other good plugin system in js ecosystem. Webpack is an interesting example because they are generally a compiler like KaTeX. I think we don't need a plugin system as sophisticated as webpack but we can use awesome ideas behind it. [You find the api here](https://webpack.js.org/api/plugins/).

**Comment by HosseinAgha** [comment]
I think we can have 2 different approaches for this:  
  1. Async: make plugins to subscribe to events (like webpack) and call all event handlers with functions that we want to expose like `defineSymbol` as their callback for example. this way we have lots of flexibility in the future but it generally have some conflicts with synchronous nature of KaTeX. In this approach we can add all plugins once before any compilations and use the changed katex object or make KaTeX to call plugin functions on every compilation to obtain symbol values for example.
  2. Sync: Add plugin changes to library before doing any compilation. This way we cannot inject plugin symbols to KaTeX symbols on every compilation. The main advantages here are that we have a slightly faster synchronous pipeline, and plugins can be a super simple objects with special values as their keys.    

       ```javascript
       import katex from 'katex';
       import katexPersianArabicPlugin from 'katex-persian-arabic-plugin';
       katex.plugin(katexPersianArabicPlugin);
       katex.render(someFormulaWithPersianStringm, element);
       ```
       ```javascript
       // a simple synchronous plugin api with no exposure to main api
       import { symbol } from 'katex-plugin-utils';
       module.exports = {
            getFontMetrics() {
                return {// object containing map of unicode number to glyph dimention }
            }
            getSymbols(createSymbol, defaults) { // createSymbol does not change anything on the main lib and is provided by `katex.plugin`
                const { groups, modes, fonts } = defaults;
            }
       }
       ```
In the synchronous approach we don't expose `defineSymbol` (and maybe a `appendFontMetrics` function!) to every plugin but we expose them to `katex.plugin` function and this function uses them to add plugins' symbols or add new fontMetrics, this way we can also have a layer of safety.

**Comment by obedm503** [comment]
I also have a similar use case @mbourne with [katex-latex ](https://github.com/obedm503/katex-latex) but I would like to have a way to intercept asciimath with some custom/specific delimiters and convert to latex before it is rendered using the `renderMathInElement` extension. so, some form of plugin system would be beneficial

**Comment by ylemkimon** [comment]
> A bigger problem is that any new function will also require code in buildHTML.js and buildMathML.js to generate the correct markup.

Wouldn't it be enough to import and use functions from main KaTeX source?

**Comment by kevinbarabash** [comment]
@ylemkimon we're working on refactoring things so that functions are defined by a single function, `defineFunction`, see https://github.com/Khan/KaTeX/blob/master/src/defineFunction.js.

**Comment by ylemkimon** [comment]
But it still seems to need functions like  `buildExpression`, `buildCommon.makeFragment`, `buildCommon.makeSpan`, etc. And I was thinking rather than exposing them to the public API, a plugin can import them from the source and be built with KaTeX.

**Comment by kevinbarabash** [comment]
@ylemkimon if they aren't exposed via the public API won't the plugin have to include all of KaTeX?

**Comment by ylemkimon** [comment]
I thought only a small portions of KaTeX is required, but it turns out that most of it is needed. Maybe use a module loader?
